crates/ra_mbe/src/syntax_bridge.rs

```diff
@@ -30,13 +59,26 @@ pub fn syntax_node_to_token_tree(node: &SyntaxNode) -> Option<(tt::Subtree, Toke
     Some((tt, token_map))
 }
 
-/// Parses the token tree (result of macro expansion) as a sequence of items
-pub fn token_tree_to_ast_item_list(tt: &tt::Subtree) -> TreeArc<ast::SourceFile> {
+/// Parses the token tree (result of macro expansion) as given macro kind
+pub fn token_tree_to_macro_kind(
+    tt: &tt::Subtree,
+    kind: MacroKind,
+) -> Result<MacroExpandResult, crate::ExpandError> {
     let token_source = SubtreeTokenSource::new(tt);
     let mut tree_sink = TtTreeSink::new(token_source.querier());
     ra_parser::parse(&token_source, &mut tree_sink);
     let syntax = tree_sink.inner.finish();
-    ast::SourceFile::cast(&syntax).unwrap().to_owned()
+
+    let res: Option<_> = match kind {
+        MacroKind::SourceFile => {
+            ast::SourceFile::cast(&syntax).map(|m| MacroExpandResult::SourceFile(m.to_owned()))
```

Comment:
The `ra_parser::parse` function should be inside this arm. `kind` should identify the parsing function we call, it should be `parse` for list of items, `expr` for expression, etc.

Let's also consult rustc to get the actual list of things macro can be parsed into: I don't think `MacroKind::Block` actually exist in the langauge?

Reply:
Here is the full list of rustc of non terminal iternoplation:
```
 NtItem(P<ast::Item>),
    NtBlock(P<ast::Block>),
    NtStmt(ast::Stmt),
    NtPat(P<ast::Pat>),
    NtExpr(P<ast::Expr>),
    NtTy(P<ast::Ty>),
    NtIdent(ast::Ident, /* is_raw */ bool),
    NtLifetime(ast::Ident),
    NtLiteral(P<ast::Expr>),
    /// Stuff inside brackets for attributes
    NtMeta(ast::MetaItem),
    NtPath(ast::Path),
    NtVis(ast::Visibility),
    NtTT(TokenTree),
    // These are not exposed to macros, but are used by quasiquote.
    NtArm(ast::Arm),
    NtImplItem(ast::ImplItem),
    NtTraitItem(ast::TraitItem),
    NtForeignItem(ast::ForeignItem),
    NtGenerics(ast::Generics),
    NtWhereClause(ast::WhereClause),
    NtArg(ast::Arg),
```
Give me some time to find where is the code to show what can be parsed into :)

Reply:
Should be : 
```
Expr(P<ast::Expr>) { "expression"; one fn visit_expr; fn visit_expr; fn make_expr; }
    Pat(P<ast::Pat>) { "pattern"; one fn visit_pat; fn visit_pat; fn make_pat; }
    Ty(P<ast::Ty>) { "type"; one fn visit_ty; fn visit_ty; fn make_ty; }
    Stmts(SmallVec<[ast::Stmt; 1]>) {
        "statement"; many fn flat_map_stmt; fn visit_stmt; fn make_stmts;
    }
    Items(SmallVec<[P<ast::Item>; 1]>) {
        "item"; many fn flat_map_item; fn visit_item; fn make_items;
    }
    TraitItems(SmallVec<[ast::TraitItem; 1]>) {
        "trait item"; many fn flat_map_trait_item; fn visit_trait_item; fn make_trait_items;
    }
    ImplItems(SmallVec<[ast::ImplItem; 1]>) {
        "impl item"; many fn flat_map_impl_item; fn visit_impl_item; fn make_impl_items;
    }
    ForeignItems(SmallVec<[ast::ForeignItem; 1]>) {
        "foreign item"; many fn flat_map_foreign_item; fn visit_foreign_item; fn make_foreign_items;
}
```

Reply:
Heh, that's larger than I've hoped for :) I wonder how rustc handles the fact that a macro can expand to multiple items, but only to a single expression...

Reply:
Aha, an error message in this case is illuminating: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=335e9a05f9629ac39c98cd61b5ac065b

Reply:
Haha, maybe you should file a bug in rustc :)
Anyway, as i remember correclty, we do not have a `ast::Stmt` type, we use `expr` , right?
```
TraitItems(SmallVec<[ast::TraitItem; 1]>) 
ImplItems(SmallVec<[ast::ImplItem; 1]>)
ForeignItems(SmallVec<[ast::ForeignItem; 1]>
```
Do we really need it ? Or we can use `Items` instead ?

Reply:
We have `ExprStmt`, which is not `Expr`. 

>Do we really need these ? Or we can use Items instead ?

Given that there are quite a few variants there, I wonder if we can start with just `Expr`, `Items` and `Stmts`? I think it makes sense to start small, figure out the required API throughout the stack, and then fill the missing cases.

Reply:
Im not so familiar about `hir` side, What result type would you think these api should return ?
Something Like these ?
```rust
pub enum MacroExpandResult {
    Expr(TreeArc<ast::Expr>),
    Pat(TreeArc<ast::Pat>),
    Ty(TreeArc<ast::Ty>),
    Stmts(Vec<TreeArc<ast::ExprStmt>>),
    Items(Vec<TreeArc<ast::Item>>),
}
```
However we do not have `ast::Item` and `ast::Ty` type, we only have `ast::AstToken` and `ast::ItemList` type.

Reply:
`ast::Item` is I think `ast::ModuleItem`, and `Ty` is a `TypeRef`.

>Something Like these ?

Hm, not exctly: it's the caller who decides what macro expands do, so it should be a bunch of functions like

```
tt_to_expr() -> TreeArc<ast::Expr>
tt_to_pat() -> TreeArc<ast::Pat>
```

For `Stmts` and `Items`, I *think* we want to return a single container node. IT probably shouldn't be `SourceFile`, we should introduce dedicated `MacroItems` and `MacroStmts`

Reply:
So `MacroItems` and `MacroStmts` would be a new ast::Node and i need to add it in `grammer.ron`, right ?

Reply:
right!

Reply:
(the whole "parse tt into SourceFile" was just a hack, the proper soln are dedicated nodes)